### ISSUE TITLE
Garesat.py

### ISSUE THREAD

**Issue by XishunZhu** [XishunZhu/Garesat_Net-SR-#5]
import numpy as np
import torch
import torch.nn as nn
import torch.nn.functional as F
import cv2
from PIL import Image
import numpy as np
import matplotlib.pyplot as plt
from PIL import Image
import os
import os.path
import torchvision.utils as vutils

def PlotImage(image):
    """
	PlotImage: Give a normalized image matrix which can be used with implot, etc.
	Maps to [0, 1]
	"""
    im = image.astype (float)
    return (im - np.min (im)) / (np.max (im) - np.min (im))
class ChannelAttention(nn.Module):
    def __init__(self, in_planes, ratio=16):
        super(ChannelAttention, self).__init__()
        self.avg_pool = nn.AdaptiveAvgPool2d(1)
        self.max_pool = nn.AdaptiveMaxPool2d(1)

        self.sharedMLP = nn.Sequential(
            nn.Conv2d(in_planes, in_planes // ratio, 1, bias=False), nn.ReLU(),
            nn.Conv2d(in_planes // ratio, in_planes, 1, bias=False))
        self.sigmoid = nn.Sigmoid()

    def forward(self, x):
        avgout=self.avg_pool (x)
        print(avgout.shape)
        avgout = self.sharedMLP(avgout)
        print(avgout.shape)
        maxout=self.max_pool(x)
        maxout = self.sharedMLP(maxout)
        return self.sigmoid(avgout + maxout)

class SpatialAttention(nn.Module):
    def __init__(self, kernel_size=7):
        super(SpatialAttention, self).__init__()
        assert kernel_size in (3,7), "kernel size must be 3 or 7"
        padding = 3 if kernel_size == 7 else 1

        self.conv = nn.Conv2d(2,1,kernel_size, padding=padding, bias=False)
        self.sigmoid = nn.Sigmoid()

    def forward(self, x):
        avgout = torch.mean(x, dim=1, keepdim=True)
        print(avgout.shape)
        maxout, _ = torch.max(x, dim=1, keepdim=True)
        print(maxout.shape)
        x = torch.cat([avgout, maxout], dim=1)
        x = self.conv(x)
        print(x.shape)
        return self.sigmoid(x)

class BasicBlock(nn.Module):
    def __init__(self,in_ch, out_ch):
        super(BasicBlock, self).__init__()
        self.sconv1 = nn.Conv2d(in_ch, out_ch, kernel_size=(3,3) ,stride=(1,1), padding=1, padding_mode='reflect')
        self.sbn1 = nn.BatchNorm2d(out_ch)
        self.srelu = nn.ReLU()

        self.sconv2 = nn.Conv2d(out_ch,out_ch,kernel_size=(3,3) ,stride=(1,1), padding=1, padding_mode='reflect')
        self.sbn2 = nn.BatchNorm2d(out_ch)
        self.srelu = nn.ReLU()

        self.ca = ChannelAttention(out_ch)
        self.sa = SpatialAttention(7)

    def forward(self, x):

        # print("B2",x.shape)
        out = self.sconv1(x)
        out = self.sbn1(out)
        out = self.srelu(out)

        out = self.sconv2(out)
        out = self.sbn2(out)
        print("B1",out.shape)
        print("SA(out)",self.sa(out).shape)
        out = self.ca(out) * out  
        out = self.sa(out) * out  
        out = self.srelu(out)
        # print("B1",out.shape)
        return out

def gaussian_2d_kernel(kernel_size=3, sigma=0):
    kernel = np.zeros ([kernel_size, kernel_size])
    center = kernel_size // 2

    if sigma == 0:
        sigma = ((kernel_size - 1) * 0.5 - 1) * 0.3 + 0.8

    s = 2 * (sigma ** 2)
    sum_val = 0
    for i in range (0, kernel_size):
        for j in range (0, kernel_size):
            x = i - center
            y = j - center
            kernel[i, j] = np.exp (-(x ** 2 + y ** 2) / s)
            sum_val += kernel[i, j]
            # /(np.pi * s)
    sum_val = 1 / sum_val
    return kernel * sum_val

class DEPTHWISECONV(nn.Module):
    def __init__(self,in_ch,out_ch):
        super(DEPTHWISECONV, self).__init__()
        self.depth_conv = nn.Conv2d(in_channels=in_ch,
                                    out_channels=in_ch,
                                    kernel_size=3,
                                    stride=1,
                                    padding=1,
                                    groups=in_ch)
        self.point_conv = nn.Conv2d(in_channels=in_ch,
                                    out_channels=out_ch,
                                    kernel_size=1,
                                    stride=1,
                                    padding=0,
                                    groups=1)
    def forward(self,input):
        out = self.depth_conv(input)
        # out = self.point_conv(out)
        return out

def SGN(x):
    x = x.detach ().numpy ()
    signResult = np.sign (x)
    signResult = torch.tensor (signResult, dtype=torch.float32, requires_grad=False)
    return signResult

def SRM(input):  

    filter1 = gaussian_2d_kernel (5, 1)
    filter2 = gaussian_2d_kernel (9, 2)
    filter3 = gaussian_2d_kernel (19, 4)
    filter1 = [[filter1]]
    filter2 = [[filter2]]
    filter3 = [[filter3]]

    filter1 = torch.tensor (filter1, dtype=torch.float32, requires_grad=False)
    filter2 = torch.tensor (filter2, dtype=torch.float32, requires_grad=False)
    filter3 = torch.tensor (filter3, dtype=torch.float32, requires_grad=False)

    # print (input.shape)
    op1 = F.conv2d (input, filter1, stride=1, padding=2)
    op2 = F.conv2d (input, filter2, stride=1, padding=4)
    op3 = F.conv2d (input, filter3, stride=1, padding=9)
    # print (op1.shape)
    # print (op2.shape)
    # print (op3.shape)

    return op1, op2, op3
class ResBlock(nn.Module):
    def __init__(self, inChannals, outChannals):
 
        super(ResBlock1, self).__init__()
        self.conv1 = nn.Conv2d(inChannals, outChannals, kernel_size=1, bias=False)
        self.bn1 = nn.BatchNorm2d(outChannals)

        self.conv2 = nn.Conv2d(outChannals, outChannals, kernel_size=3, stride=1, padding=1, bias=False)
        self.bn2 = nn.BatchNorm2d(outChannals)

        self.conv3 = nn.Conv3d(outChannals*2, outChannals, kernel_size=3, stride=1, padding=1, bias=False)
        self.bn3 = nn.BatchNorm2d(outChannals)

        self.conv4 = nn.Conv2d(outChannals*2, outChannals, kernel_size=1, bias=False)
        self.bn4 = nn.BatchNorm2d(outChannals)
        self.relu = nn.ReLU(inplace= False)
        self.BasicBlock = self._makeLayer_(BasicBlock, out_ch , out_ch, 1)

    def _makeLayer_(self, block, inChannals, outChannals, blocks):

        layers = []
        layers.append(block(inChannals, outChannals))

        for i in range(1, blocks):
            layers.append(block(outChannals, outChannals))

        return nn.Sequential(*layers)
    def forward(self, x)
        resudial = x
        out = self.conv1(x)
        out = self.bn1(out)
        out = self.relu(out)
        Aout= self.BasicBlock(out)
        Aout = self.conv2(Aout)
        Aout = self.bn2(Aout)
        Aout =self.relu(Aout)

        out1 = torch.empty([1, 64, out.shape[2], out.shape[3]])
        out2 = torch.empty ([1, 64, out.shape[2], out.shape[3]])
        out3 = torch.empty ([1, 64, out.shape[2], out.shape[3]])
        for i in range(out.shape[1]):
            y = out[:,i:i+1,::]
            y1,y2,y3 = SRM(y)
            out1[:,i:i+1,::]=y1
            out2[:, i:i + 1, ::] = y2
            out3[:, i:i + 1, ::] = y3

        D1= out-out1
        D2 = out1 - out2
        D3 = out2 - out3
        D_1= out+(1-0.25*SGN(D1))*D1+0.5*D2+0.5*D3

        D = self.conv2(D_1)
        D = self.bn2(D)
        D = self.relu(D)

        D = self.conv3(D)
        D = self.bn2(D)
        D = self.relu(D)
        AGout = torch.cat([D,Aout],dim=1)
        AGout = self.conv3(AGout)
        AGout = self.bn3(AGout)
        AGout = self.relu(AGout)
        FGout = torch.cat([out, AGout], dim=1)
        FGout = self.conv4(FGout)
        FGout = self.bn4(FGout)
        FGout = self.relu(FGout)
        return FGout

def show_feature_map(feature_map):
    print("1",feature_map.shape)
    # feature_map = feature_map.unsqueeze(1)  #
    # print ("2",feature_map.shape)
  
    # feature_map = feature_map.view(1, feature_map.shape[1], feature_map.shape[2], feature_map.shape[3])  # (1,64,55,55)
    print(feature_map.shape)
    upsample = torch.nn.UpsamplingBilinear2d(size=(feature_map.shape[2], feature_map.shape[3]))  
    feature_map = upsample(feature_map)
    feature_map = feature_map.view(feature_map.shape[1], feature_map.shape[2], feature_map.shape[3])
    print(feature_map.shape)
    feature_map_num = feature_map.shape[0] 
    row_num = np.ceil(np.sqrt(feature_map_num))  
    plt.figure()
    for index in range(1, feature_map_num + 1):

        plt.subplot(row_num, row_num, index)
        print(feature_map[index - 1].shape)
        #plt.imshow(transforms.ToPILImage()(feature_map[index - 1]).convert('RGB'))
        plt.imshow(feature_map[index - 1], cmap='gray')  # feature_map[0].shape=torch.Size([55, 55])
       1]))#feature_map[0].shape=torch.Size([55, 55])
        plt.axis('off')
        plt.imsave(
            r'E:\Xishun\supercostruction\training\featuremap'  + str(index) + ".png",
            feature_map[index - 1])
    plt.show()

class Garesat_Net(nn.Module):
    def __init__(self, input_nc=3,out_ch=64):
        super(Garesat_Net, self).__init__()

        self.tconv1 = nn.Conv2d(input_nc, out_ch,kernel_size=(3,3), stride=(1,1), padding=1,padding_mode='reflect')
        self.tbn1 = nn.BatchNorm2d(out_ch)

        self.tconv2 = nn.Conv2d(out_ch, out_ch, kernel_size=(3,3), stride=(1,1), padding=1,padding_mode='reflect')
        self.tbn2 = nn.BatchNorm2d(out_ch)

        self.tconv3 = nn.Conv2d(out_ch, out_ch, kernel_size=(5,5), stride=(1,1), padding=2,padding_mode='reflect')
        self.tbn3 = nn.BatchNorm2d(out_ch)

        self.tconv4 = nn.Conv2d(out_ch, out_ch, kernel_size=(3,3), stride=(1,1), padding=1)

        self.tbn4 = nn.BatchNorm2d(out_ch)
        self.tconv5 = nn.Conv2d(out_ch, out_ch*4,kernel_size=(5,5), stride=(1,1), padding=2)
        self.tbn5 = nn.BatchNorm2d(out_ch*4)
        self.pixelShuffler = nn.PixelShuffle(2)
        self.tconv6 = nn.Conv2d(out_ch*2, out_ch, kernel_size=(5,5), stride=(1,1), padding=2)
        self.tbn6 = nn.BatchNorm2d(out_ch)
        #
        self.tconv7 = nn.Conv2d(out_ch*4, out_ch*8, kernel_size=(3,3), stride=(1,1), padding=1)
        self.tbn7 = nn.BatchNorm2d(out_ch*8)

        self.tconv8 = nn.Conv2d(out_ch*8, out_ch*8, kernel_size=(3,3), stride=(1,1), padding=1)
        self.tbn8 = nn.BatchNorm2d(out_ch*8)

        self.tconv9 = nn.Conv2d(out_ch*8, out_ch*16, kernel_size=(3,3), stride=(1,1), padding=1)
        self.tbn9 = nn.BatchNorm2d(out_ch*16)

        self.tconv10 = nn.Conv2d(out_ch, out_ch, kernel_size=(3,3), stride=(1,1), padding=1)
        self.tbn10 = nn.BatchNorm2d(out_ch)

        self.tconv11 = nn.Conv2d(out_ch, 3, kernel_size=(3,3), stride=(1,1), padding=1) #16*16
        self.tbn11 = nn.BatchNorm2d(3)

        self.resBlock = self._makeLayer_(ResBlock, out_ch, out_ch, 1)

        self.tanh = nn.Tanh()
        self.relu = nn.ReLU()
        self.drop_path_prab =0.0

    def _makeLayer_(self, block, inChannals, outChannals, blocks):
        """resblock layer"""
        layers = []
        layers.append(block(inChannals, outChannals))

        for i in range(1, blocks):
            layers.append(block(outChannals, outChannals))

        return nn.Sequential(*layers)

    def forward(self, x):

        y = self.tconv1(x)
        y = self.tbn1(y)
        y = self.relu(y)

        y = self.tconv2(y)
        y = self.tbn2(y)
        y = self.relu(y)
        y=self.resBlock(y)

        y = self.tconv3(y)
        y = self.tbn3(y)
        y = self.relu(y)

        y = self.resBlock(y)

        y = self.tconv4(y)
        y = self.tbn4(y)
        y = self.relu(y)

        y = self.tconv5(y)
        y = self.pixelShuffler(y)
        y = self.tbn5(y)
        y = self.relu(y)

        y = self.tconv6(y)
        y = self.tbn6(y)
        y = self.relu(y)

        y = self.tconv7(y)
        y = self.tbn7(y)
        y = self.relu(y)

        y = self.tconv8(y)
        y = self.tbn8(y)
        y = self.relu(y)
        y = self.resBlock(y)

        y = self.tconv9(y)
        y = self.tbn9(y)
        y = self.relu(y)

        y = self.tconv10(y)
        y = self.tbn10(y)
        y = self.relu(y)

        y = self.tconv11(y)
        y = self.tbn11(y)
        y = self.relu(y)
        return  y
